### ISSUE TITLE
Expansion of @ tokens suffers from ID collisions

### ISSUE THREAD

**Issue by AJChapman** [ivcap-works/ivcap-cli#24]
The expansion of @ tokens (e.g. `@1` into `urn:ivcap:account:fbbddd13-c9a3-472f-a776-d4ffd6c7b9be`) happens differently depending on your request history. For example, if you run `ivcap service list` and `ivcap artifact list`, and there exists both a service and an artifact with token `@1`, then in a subsequent command `@1` will expand to whichever was run last:

![image](https://github.com/ivcap-works/ivcap-cli/assets/8316672/4a9da728-149c-44fa-850c-ced43caa5520)

Note the different error messages depending on which list was requested last.

The cause of this is that the CLI saves the history of @ expansions in `~/.config/ivcap/history.yaml`, but doesn't save them by resource type, only by @ token. See `func GetHistory` in `cmd/root.go`.

**Comment by maxott** [comment]
Yes, that's how it was designed. I agree that it would be often more convenient to do what you suggest, but to further avoid confusion, we would very likely have to expand the history token to also include an indication of type, such as `@s1`, `@o1`, ... as having two o`@1` in a command, both referring to different resources may be rather confusing. Unfortunately, we now have _(a)artifact_ and _(a)aspect_.

Another approach would be to a) reuse the history token if it's in the history and b) to use a largish (~99) history 'ring'. Meaning we reuse old tokens when we have used up all available tokens.

Interested to hear what folks think. What other choices do we have?

**Comment by AJChapman** [comment]
Can the CLI always know what type the thing it's expanding /should/ have? I.e. in my `order create` example above, it knows that the first `@1` should be a service, and can probably also infer that the second one should be an artifact. So `history.yaml` can contain maps with pairs of strings as keys, e.g. ("artifact", "@1"): "urn:ivcap:artifact:...".

**Comment by AJChapman** [comment]
Another option is we don't use numbered tokens, and instead do what `git` does, with truncations of the hashes. E.g. `urn:ivcap:account:fbbdd13-c9a3-472f-aa776-d4ffd6c7b9be` becomes `7b9be`.

**Comment by asherbender** [comment]
> Another option is we don't use numbered tokens, and instead do what `git` does, with truncations of the hashes. E.g. `urn:ivcap:account:fbbdd13-c9a3-472f-aa776-d4ffd6c7b9be` becomes `7b9be`.

 I was thinking this as well... I didnt know how hard it would be to implement... so I didnt mention it...
 
 It would be handy to  type in the first few characters of the hexidecimal hash and either <tab> complete it... or trust that you typed in an adequate number of characters to define a unique hash (the CLI could throw an error if you didnt type in enough characters for it to be unique)
 
I also think the expansion of @ tokens is neat... it is easier to look up and read a number than a hexidecimal hash (of 5-6 characters). Linux shells offer a similar concept using the ! character. From the bash man page:

```
The two or three characters which control history expansion  and  tokenization  (see
HISTORY  EXPANSION  below).  The first character is the history expansion character,
the character which signals the start of a history  expansion,  normally  `!'. 
```
 
This is effectively what max has implemented... You can invoke an old command by finding its position in the history and issuing `!<number>`. Many (most?) shells support it... it is a cool feature if you arent using it already.

If we had infinite resources... it would be cool to implement both :-P

**Comment by maxott** [comment]
> It would be handy to type in the first few characters of the hexidecimal hash and either complete it... or trust that you typed in an adequate number of characters to define a unique hash (the CLI could throw an error if you didnt type in enough characters for it to be unique)

I played around with that for a bit but I couldn't get the auto-completion working like that in bash. 

>Can the CLI always know what type the thing it's expanding /should/ have? 

That is an interesting thought. It definitely should work for the 'primary' entities - service, order, artifact. Need to check if there are issues with 'aspect'. Still not sure how confusing having two identical @x in a single command where each one refers to something different.

> Another option is we don't use numbered tokens, and instead do what git does, with truncations of the hashes.

How exactly would that work? Would we display the full urn, but then can use the abbreviated version in future calls? Also, remember not everything follows the `urn:...:_uuid_` format. 

> I also think the expansion of @ tokens is neat...
What exactly do you mean by that - expansion ?

**Comment by asherbender** [comment]
> Still not sure how confusing having two identical @x in a single command where each one refers to something different.

I think that is a valid critique. 

A user who understands the interface could learn to keep that context in their head... but it isn't immediately obvious. Could you make that difference explicit by prepending each @x with the 'primary entity'?

So the following:

`ivcap order create -n "test image order" @1 msg ="Hello IVCAP" background-img=@1`

becomes something like:

`ivcap order create -n "test image order" service@1 msg ="Hello IVCAP" background-img=artifact@1`

**Comment by AJChapman** [comment]
I like @asherbender's suggestion to add the type to the @ syntax. Could we combine this with type inference where available, so that adding the type is optional? The user could use these interchangeably:

```
ivcap order create -n "foo" @service:1 msg="Hello IVCAP" background-img=@artifact:1
```

```
ivcap order create -n "foo" @1 msg="Hello IVCAP" background-img=@1
```

If the type inference doesn't work then the CLI can suggest the user disambiguate with the full @ syntax.
